Added feature/paging

SpeckiJ opened the pull request:
Implementation of basic Query-Parameter based Paging. 

Calculation of last Page is broken! WIP

SpeckiJ committed:
Added basic Paging Functionality using supplied Query Parameters 'limit' and 'offset'.

The Paging is implemented as a wrapper around the OutputCollection Class.

Calculation of last Element is broken! WIP

further Documentation is to be found in the series-rest-api official documentation.

SpeckiJ committed:
fixed License Header.

ridoo commented on rest/src/main/java/org/n52/web/ctrl/ParameterController.java:
```diff
@@ -132,22 +135,48 @@ public void getRawData(HttpServletResponse response,
+        // Add Paging if requested
+        if (queryMap.containsParameter("limit") || queryMap.containsParameter("offset")){
+            Paginated<T> paginated = new Paginated(new OffsetBasedPagination(queryMap.getOffset(), queryMap.getLimit()), result.size());
+            this.addPagingHeaders(request.getRequestURI(), response, paginated);
```
here you should be careful, as `request.getRequestURI()` does not give you the /real/ URI the client might have called (think of a proxied environment). The external URL (configured in `application.properties`) can be obtained via `IoParameters.map.getHrefBase()`. You'll then have no need anymore for the `HttpServletRequest` parameter :).

However, you've to know the actual collection to create the right link, .e.g `http://localhost:8080/webapp-context/api/datasets` rather. For this, you could add an abstract method `getCollectionPath()` all parameter controllers have to implement and construct the full URL via [UrlHelper#constructHref(String, String)](https://github.com/52North/series-rest-api/blob/develop/rest/src/main/java/org/n52/web/ctrl/UrlHelper.java#L69)

SpeckiJ committed:
fixed whitespaces to comply with mvn tests.

ridoo commented on rest/src/main/java/org/n52/web/ctrl/ParameterController.java:
```diff
@@ -132,22 +135,48 @@ public void getRawData(HttpServletResponse response,
+        }
+        
+        // Add Paging if requested
+        if (queryMap.containsParameter("limit") || queryMap.containsParameter("offset")){
```
could you please move paging to [ParameterRequestMappingAdapter](https://github.com/52North/series-rest-api/blob/develop/rest/src/main/java/org/n52/web/ctrl/ParameterRequestMappingAdapter.java#L52) .. I think it is better placed here as it is enhancing and auto handling requests

SpeckiJ committed:
Refactored Paging from `ParameterController` to `ParameterRequestMappingadapter`.

fixed Calculation of last Elements.
fixed External URL Aquisation.

ServicesParameterController & GeometriesController still WIP! Paging currently disabled for both.

SpeckiJ commented:
TODO:

SpeckiJ committed:
Merge remote-tracking branch 'upstream/develop' into feature/paging

SpeckiJ committed:
Fixed NullPointerException on using limit and offset filters.

SpeckiJ committed:
fixed page calculations.

self reference is always present if offset is valid
first reference is always present if offset is valid
last reference is always present if offset is valid

next reference is present if exists and offset is valid
previous reference is present if exists and offset is valid

SpeckiJ committed:
removed limit and offset to prevent `ORDER BY` in SQL `count(*)` which is incompatible with Hibernate.

SpeckiJ committed:
removed unused import.

Implemented Paging for Services. Needs getServiceCount() Functionality in doa-series-api.

SpeckiJ commented:
TODO: 

Update Documentation

phenomena: wrong numbers ?

ridoo commented on rest/pom.xml:
```diff
@@ -18,6 +18,16 @@
+        </dependency>
+        <dependency>
+            <groupId>org.n52.series-api.db</groupId>
+            <artifactId>dao</artifactId>
```
don' mix SPI implementation with SPI, i.e. in every case the SPI is independent from its implementations (otherwise all implementations depend on each other)

ridoo commented on rest/src/main/java/org/n52/web/ctrl/CategoriesParameterController.java:
```diff
@@ -29,12 +29,25 @@
+
+    @Override
+    protected int getElementCount(IoParameters queryMap) throws DataAccessException {
+        return super.getEntityCounter().countCategories(new DbQuery(queryMap));
```
don't use the `DbQuery` class from the dao module as it is an implementation of the SPI and specific to database access.

ridoo commented on rest/src/main/java/org/n52/web/ctrl/ParameterRequestMappingAdapter.java:
```diff
@@ -52,13 +63,29 @@
+                queryMap = queryMap.removeAllOf("limit").removeAllOf("offset");
+                Paginated<T> paginated = new Paginated(obp, this.getElementCount(queryMap));
+                this.addPagingHeaders(this.getCollectionPath(queryMap.getHrefBase()), response, paginated);
+            } catch (DataAccessException ex) {
```
As above .. `DataAccessException` shouldn't be handled in the Web controller layer as it's specific to the DAO SPI implementation

ridoo reviewed (changes requested):
Check dependency and use those types available from the SPI (`QueryParameters` or `IoParameters`)

SpeckiJ committed:
refactored to use EntityCounterWrapper to remove dependency on dao implementation.

SpeckiJ committed:
mvn checkstyle compliance

SpeckiJ committed:
Refactored counter to exclude specific DAO Implementations

SpeckiJ committed:
mvn checkstyle compliance once again.

SpeckiJ commented:
TODO:

Documentation
Check if EntityCounter Numbers are correct
StationsParameterController Paging -> missing externalURL
TimeseriesDataController Paging -> missing externalURL

SpeckiJ committed:
fixed incorrect count on caused by backwards compatibility.
mvn checkstyle compliance.

SpeckiJ committed:
Added Paging for /stations Endpoint.

SpeckiJ committed:
Added basic documentation about paging.

SpeckiJ committed:
fixed self-referencing dependency.

SpeckiJ commented:
Moved PR to official Repository. New PR is [Added Paging Functionality based on query parameters. #386]https://github.com/52North/series-rest-api/pull/386).
